### PR TITLE
Scope and associate user's study to their session

### PR DESCRIPTION
**Pull request by alxdarksage**
An important step towards study-scoped authentication:
- Fixes security hole where you could sign in under one study, then act under another one by using a different study identifier in the header/host. Study information now comes from a user session. Once authenticated, that's the study you're operating under.
- Stop using study when you're only looking for the identifier or researcher role, but use StudyIdentifier to provide type safety. No trip to Redis or DynamoDB.
- Only look for study in the header, host name or configuration file when doing an action like signing in, where we don't have an authenticated session.
- v2 authentication will explicitly ask for the study you are signing in to, and go to that study's directory to authenticate you.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSchedulePlanDao.java [hunk 1]
```diff
@@ -53,13 +53,13 @@ public void setDynamoDbClient(AmazonDynamoDB client) {
     }
     
     @Override
-    public SchedulePlan getSchedulePlan(Study study, String guid) {
-        checkNotNull(study, "Study is null");
-        checkArgument(StringUtils.isNotBlank(study.getIdentifier()), "Study key is null");
+    public SchedulePlan getSchedulePlan(StudyIdentifier studyIdentifier, String guid) {
+        checkNotNull(studyIdentifier, "studyIdentifier is null");
+        checkArgument(StringUtils.isNotBlank(studyIdentifier.getIdentifier()), "Study key is null");
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoSchedulePlanDao.java]: this is now impossible so these can be removed.

File: test/controllers/AuthenticationControllerTest.java [hunk 2]
```diff
@@ -90,4 +106,35 @@ public void testCode() throws Exception {
         });
     }
     
+    @Test
+    public void onceAuthenticatedUserCannotSwitchStudies() {
```
Comment (alxdarksage) [inline test/controllers/AuthenticationControllerTest.java:115]: This could be an SDK test.

File: app/controllers/BaseController.java [hunk 3]
```diff
@@ -64,17 +68,25 @@ public void setBridgeConfig(BridgeConfig bridgeConfig) {
     }
     
     /**
+     * Returns a session. Will not throw exception if user is not authorized or has not consented to research.
+     * @return session if it exists, or null otherwise.
+     */
+    protected UserSession getSessionIfItExists() {
+        String sessionToken = getSessionToken();
+        return authenticationService.getSession(sessionToken);
```
Comment (DwayneJengSage) [inline app/controllers/BaseController.java:79]: What happens if you call getSession with a null or empty session token? Should have have a check here? Or possibly in authService?

File: test/controllers/AuthenticationControllerTest.java [hunk 4]
```diff
@@ -45,19 +57,49 @@
...
+    
     @Before
     public void before() {
         testUser = helper.createUser(AuthenticationControllerTest.class);
+        
+        String id = RandomStringUtils.randomAlphabetic(7).toLowerCase();
+        secondStudy = new DynamoStudy();
+        secondStudy.setIdentifier(id);
+        secondStudy.setName("Second Test Study");
+        studyService.createStudy(secondStudy);
```
Comment (DwayneJengSage) [inline test/controllers/AuthenticationControllerTest.java]: @Before is run with every test. Not every test needs the second study. You should either change this to @BeforeClass or move it into the individual test method.

**Pull request merged**